src/utils/stl_parser.rs

```diff
@@ -0,0 +1,267 @@
+
+impl STLParser {
+    pub fn read_stl(path: PathBuf) -> anyhow::Result<Triangles> {
+        match STLParser::read_binary_stl(path.clone()) {
+            Ok(triangles) => Ok(triangles),
+            Err(err_bin) => match STLParser::read_ascii_stl(path) {
+                Ok(triangles) => Ok(triangles),
+                Err(err) => {
+                    Err(anyhow::anyhow!(
+                    "Failed to read STL file in either binary or ascii format!\nErrors:\n{}\n{}",
+                    err_bin,
+                    err
+                ))
+                }
+            },
+        }
+    }
+
+    fn read_binary_stl(path: PathBuf) -> anyhow::Result<Triangles> {
+        let mut file = File::open(path.clone())
+            .with_context(|| format!("Failed to open file \'{}\'!", path.display()))?;
+        let mut triangles = Triangles::new();
+        let mut header = [0u8; 80];
+        file.read_exact(&mut header)
+            .with_context(|| format!("Failed to read header from binary STL!"))?;
+
+        let mut num_triangles_bytes = [0u8; 4];
+        file.read_exact(&mut num_triangles_bytes)
+            .with_context(|| format!("Failed to read number of triangles from binary STL!"))?;
+
+        let num_triangles = u32::from_le_bytes(num_triangles_bytes);
+
+        for _i in 0..num_triangles {
+            let mut buffer = [0u8; 50];
+            file.read_exact(&mut buffer)
+                .with_context(|| format!("Failed to read triangle from binary STL!"))?;
+
+            let normal: Vec3 = (
+                f32::from_le_bytes([buffer[0], buffer[1], buffer[2], buffer[3]]),
+                f32::from_le_bytes([buffer[4], buffer[5], buffer[6], buffer[7]]),
+                f32::from_le_bytes([buffer[8], buffer[9], buffer[10], buffer[11]]),
+            )
+                .into();
+
+            let mut vertices = [[0.0; 3]; 3];
+            for i in 0..3 {
+                let offset = 12 + i * 12;
+                vertices[i][0] = f32::from_le_bytes([
+                    buffer[offset],
+                    buffer[offset + 1],
+                    buffer[offset + 2],
+                    buffer[offset + 3],
+                ]);
+                vertices[i][1] = f32::from_le_bytes([
+                    buffer[offset + 4],
+                    buffer[offset + 5],
+                    buffer[offset + 6],
+                    buffer[offset + 7],
+                ]);
+                vertices[i][2] = f32::from_le_bytes([
+                    buffer[offset + 8],
+                    buffer[offset + 9],
+                    buffer[offset + 10],
+                    buffer[offset + 11],
+                ]);
+            }
+
+            triangles.push(Triangle {
+                normal,
+                vertices: [vertices[0].into(), vertices[1].into(), vertices[2].into()],
+            })
+        }
+
+        Ok(triangles)
+    }
+
+    pub fn write_binary_stl(path: PathBuf, triangles: Triangles) -> anyhow::Result<()> {
```

Comment:
if you are writing to a file, then this function should be named "_write_to_file_".
otherwise, when you name your function without a "__file_" postfix, then you should abstract your code using the `Write` trait or smth like that

Reply:
this comment also applies to all functions that work with files